core/http-auth/src/main/java/software/amazon/awssdk/http/auth/internal/DefaultAwsV4HttpSigner.java

```diff
@@ -72,157 +75,225 @@ public SyncSignedRequest sign(SyncSignRequest<? extends AwsCredentialsIdentity>
+            throw new IllegalArgumentException(
+                CHECKSUM_ALGORITHM + " cannot be null when " + CHECKSUM_HEADER_NAME + " is given!");
         }
-        requestBuilder.putHeader(SignerConstant.HOST, host);
-    }
 
-    private void addDateHeader(SdkHttpRequest.Builder requestBuilder, String dateTime) {
-        requestBuilder.putHeader(SignerConstant.X_AMZ_DATE, dateTime);
+        return createSdkChecksumFromRequest(request.request(), checksumHeaderName, checksumAlgorithm);
     }
 
-    private void putChecksumHeader(SdkChecksum sdkChecksum,
-                                   SdkHttpRequest.Builder requestBuilder, String contentHashString, String checksumHeaderName) {
+    /**
+     * Generate a {@link ContentChecksum} from the {@link SyncSignRequest}.
+     */
+    protected ContentChecksum createChecksum(SyncSignRequest<?> request) {
+        SdkChecksum sdkChecksum = createSdkChecksum(request);
+        String contentHash = createContentHash(request.payload().orElse(null), sdkChecksum);
 
-        if (sdkChecksum != null && !isUnsignedPayload(contentHashString)) {
-            requestBuilder.putHeader(checksumHeaderName, BinaryUtils.toBase64(sdkChecksum.getChecksumBytes()));
-        }
+        return new ContentChecksum(contentHash, sdkChecksum);
     }
 
-    private Boolean isUnsignedPayload(String contentHashString) {
-        return "UNSIGNED_PAYLOAD".equals(contentHashString) || "STREAMING-UNSIGNED-PAYLOAD-TRAILER".equals(contentHashString);
+    /**
+     * Generate a {@link CompletableFuture<ContentChecksum>} from the {@link AsyncSignRequest}
+     */
+    protected CompletableFuture<ContentChecksum> createChecksum(AsyncSignRequest<?> request) {
+        SdkChecksum sdkChecksum = createSdkChecksum(request);
+
+        return createContentHash(request.payload().orElse(null), sdkChecksum).thenApply(
+            hash -> new ContentChecksum(hash, sdkChecksum));
     }
 
-    @Override
-    public AsyncSignedRequest signAsync(AsyncSignRequest<? extends AwsCredentialsIdentity> request) {
-        if (CredentialUtils.isAnonymous(request.identity())) {
-            return AsyncSignedRequest.builder()
-                .request(request.request())
-                .payload(request.payload().orElse(null))
-                .build();
-        }
+    /**
+     * Add any prerequisite items to the request using the {@link SdkHttpRequest.Builder}, the ${@link SignRequest},
+     * and the {@link ContentChecksum}
+     * <p>
+     * Such an item could be a header or query parameter that should be included in the signature of the request.
+     */
+    protected void addPrerequisites(SdkHttpRequest.Builder requestBuilder,
+                                    SignRequest<?, ? extends AwsCredentialsIdentity> signRequest,
+                                    ContentChecksum contentChecksum) {
+        Instant requestSigningInstant = validatedProperty(signRequest, SIGNING_CLOCK).instant();
```

Comment:
similar bug alert here too. This could mean the instant used in the x-amz-date header would be different than the one used in the scope.

Reply:
I'm thinking how this should have been caught in testing. Not sure if doable, but I think look at if there is a way to setup the Clock in the test such that it uses the fixed time the first time you call instant(), but make the clock tick such that any subsequent call to instant gives a different time, ideally different date, so we explicitly set the test up like that to prevent potential regressions in the future too.

Reply:
I can add a test to verify.

Reply:
I didn't see a way to instantiate a Clock with the java.time.Clock methods that'd allow this. Probably good to use a mock Clock with `when(mockClock.instant()).thenReturn(instant).thenReturn(instantPlusADay)`. And possibly also asserting using `verify` with `times(1)` or `verifyNoMore`.

Reply:
I've done it in the event-stream test, I simply created an implementation of Clock that allows for setting the time.